File: src/main/java/components/SpriteRenderer.java
package components;

import jade.Component;
import jade.Transform;
import org.joml.Vector2f;
import org.joml.Vector4f;
import renderer.Texture;

public class SpriteRenderer extends Component {

    private Vector4f color;
    private Sprite sprite;
    private Transform lastTransform;
    private boolean isDirty = false;

    public SpriteRenderer(Vector4f color) {
        this.color = color;
        this.sprite = new Sprite((Texture) null);
        isDirty = true;
    }

    public boolean isDirty() {
        return isDirty;
    }

    public void setDirty(boolean dirty) {
        isDirty = dirty;
    }

    public void setClean() {
        this.setDirty(false);
    }

    public SpriteRenderer(Sprite sprite) {
        this.sprite = sprite;
        this.color = new Vector4f(1, 1, 1, 1);
        isDirty = true;
    }

    @Override
    public void start() {
        super.start();
        lastTransform = gameObject.transform.copy();
    }

    @Override
    public void update(float dt) {
        if (!lastTransform.equals(gameObject.transform)) {
            gameObject.transform.copy(lastTransform);
            isDirty = true;
        }
    }

    public Vector4f getColor() {
        return color;
    }

    public Vector2f[] getTexCoords() {
        return sprite.getTexCoords();
    }

    public Texture getTexture() {
        return sprite.getTexture();
    }

    public void setColor(Vector4f color) {
        if (!this.color.equals(color)) {
            this.color.set(color);
            isDirty = true;
        }
    }

    public void setSprite(Sprite sprite) {
        this.sprite = sprite;
        isDirty = true;
    }
}

File: src/main/java/components/SpriteRenderer.kt
package components


import imgui.ImGui
import jade.Component
import jade.Transform
import org.joml.Vector2f
import org.joml.Vector4f
import renderer.Texture

class SpriteRenderer : Component {
    var color: Vector4f
        set(color) {
            if (this.color != color) {
                field = color
                isDirty = true
            }

        }
    private var sprite: Sprite? = null
    private lateinit var lastTransform: Transform
    var isDirty = false

    constructor(color: Vector4f) {
        this.color = color
//        sprite = Sprite((null as Texture?)!!)
        isDirty = true
    }

    fun setClean() {
        isDirty = false
    }

    constructor(sprite: Sprite) {
        this.sprite = sprite
        color = Vector4f(1F, 1F, 1F, 1F)
        isDirty = true
    }

    override fun start() {
        super.start()
        lastTransform = gameObject.transform.copy()
    }

    override fun update(dt: Float) {
        if (lastTransform != gameObject.transform) {
            gameObject.transform.copy(lastTransform)
            isDirty = true
        }
    }

    override fun imgui() {
        val imColor = floatArrayOf(color.x, color.y, color.z, color.w)
        if (ImGui.colorPicker4("Color Picker: ", imColor)) {
            color.set(imColor[0], imColor[1], imColor[2], imColor[3])
            isDirty = true
        }

    }


    val texCoords: Array<Vector2f>?
        get() = sprite?.texCoords
    val texture: Texture?
        get() = sprite?.texture


    fun setSprite(sprite: Sprite) {
        this.sprite = sprite
        isDirty = true
    }
}